pkg/operator/ceph/disruption/machinelabel/reconcile.go

```diff
@@ -0,0 +1,140 @@
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	controllerName                  = "machinelabel-controller"
+	MachineFencingLabelKey          = "fencegroup.rook.io/cluster"
+	MachineFencingNamespaceLabelKey = "fencegroup.rook.io/clusterNamespace"
+	OSDPodLabelValue                = "rook-ceph-osd"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", controllerName)
+
+type ReconcileMachineLabel struct {
+	scheme  *runtime.Scheme
+	client  client.Client
+	options *controllerconfig.Context
+}
+
+type machine struct {
+	isOccupiedByOSD bool
+	RawMachine      mapiv1.Machine
+}
+
+// Reconcile is the implementation of reconcile function for ReconcileMachineLabel
+// which ensures that the machineLabel for the osd pods are in correct state
+func (r *ReconcileMachineLabel) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger.Infof("reconciling %s", request.NamespacedName)
+
+	// Fetch list of osd pods for the requested ceph cluster
+	pods := &corev1.PodList{}
+	err := r.client.List(context.TODO(), pods, client.InNamespace(request.Namespace), client.MatchingLabels{k8sutil.AppAttr: OSDPodLabelValue, k8sutil.ClusterAttr: request.Name})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Fetching the cephCluster
+	cephClusterInstance := &cephv1.CephCluster{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, cephClusterInstance)
+	if err != nil {
+		logger.Errorf("failed to fetch cephcluster %s: %+v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	// skipping the reconcile since the feature is switched off
+	if !cephClusterInstance.Spec.DisruptionManagement.ManageMachineDisruptionBudgets {
+		logger.Debugf("Skipping reconcile for cephCluster %s as manageMachineDisruption is turned off", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch list of machines available
+	machines := &mapiv1.MachineList{}
+	err = r.client.List(context.TODO(), machines, client.InNamespace(cephClusterInstance.Spec.DisruptionManagement.MachineDisruptionBudgetNamespace))
+	if err != nil {
+		logger.Errorf("failed tp fetch machine list %+v", machines)
+		return reconcile.Result{}, err
+	}
+
+	nodeMachineMap := map[string]machine{}
+
+	// Adding machines to nodeMachineMap
+	for _, m := range machines.Items {
+		if m.Status.NodeRef != nil {
+			nodeMachineMap[m.Status.NodeRef.Name] = machine{RawMachine: m}
+		}
+	}
+
+	// Marking machines that are occupied by the osd pods
+	for _, pod := range pods.Items {
+		if pod.Spec.NodeName != "" {
+			if machine, p := nodeMachineMap[pod.Spec.NodeName]; p {
+				machine.isOccupiedByOSD = true
+				nodeMachineMap[pod.Spec.NodeName] = machine
+			}
+		}
+	}
+
+	// Updating the machine status
+	for _, machine := range nodeMachineMap {
```

Comment:
How often would this reconciler be triggered in a typical cluster? Is it only be triggered when a node is fenced? If it is more often such as anytime a node is updated (which happens quite often in another node controller in rook), it could be a lot of machine updates.

Reply:
If it is very frequent, we should only log when something changes, not every reconcile.

Reply:
ok changed.